### PR TITLE
Add Migrate and abortMigration actions

### PR DESCRIPTION
**Pull request by gitlawr**
1. Introduce VMIM types
2. enable LiveMigration feature gates by default
3. Add Migrate/abortMigration action
#133

### CODE REVIEW

File: pkg/api/vm/formatter.go [hunk 1]
```diff
@@ -124,6 +134,20 @@ func (vf *vmformatter) canStop(vmi *v1alpha3.VirtualMachineInstance) bool {
 	return true
 }
 
+func (vf *vmformatter) canMigrate(vmi *v1alpha3.VirtualMachineInstance) bool {
+	if vmi.IsRunning() && (vmi.Status.MigrationState == nil || vmi.Status.MigrationState.Completed) {
```
Comment (guangbochen) [inline pkg/api/vm/formatter.go]: need to add `vmi != nil` validation

File: pkg/api/vm/formatter.go [hunk 2]
```diff
@@ -124,6 +134,20 @@ func (vf *vmformatter) canStop(vmi *v1alpha3.VirtualMachineInstance) bool {
...
+		return true
+	}
+	return false
+}
+
+func (vf *vmformatter) canAbortMigrate(vmi *v1alpha3.VirtualMachineInstance) bool {
+	if vmi.Status.MigrationState != nil && !vmi.Status.MigrationState.Completed {
```
Comment (guangbochen) [inline pkg/api/vm/formatter.go]: ditto

**Review by guangbochen — COMMENTED** [review]
the backend design looks fine to me, the only thing we need to discuss is the UI design and its process, let's post it to the https://github.com/rancher/harvester/issues/133

**Review by guangbochen — APPROVED** [review]
LGTM, thanks

File: pkg/api/vm/handler.go [hunk 3]
```diff
@@ -27,54 +30,65 @@ type vmActionHandler struct {
...
 		}
 
 		if err := h.ejectCdRom(name, namespace, input.DiskNames); err != nil {
-			rw.WriteHeader(http.StatusInternalServerError)
-			rw.Write([]byte(err.Error()))
+			return err
+		}
+	case migrate:
+		if err := h.migrate(namespace, name); err != nil {
+			return err
```
Comment (aiwantaozi) [inline pkg/api/vm/handler.go]: can we directly ```return h.migrate(namespace, name)```

File: pkg/api/vm/handler.go [hunk 4]
```diff
@@ -27,54 +30,65 @@ type vmActionHandler struct {
...
 		}
 
 		if len(input.DiskNames) == 0 {
-			rw.WriteHeader(http.StatusBadRequest)
-			rw.Write([]byte("Parameter diskNames is empty"))
-			return
+			return apierror.NewAPIError(validation.InvalidBodyContent, "Parameter diskNames is empty")
 		}
 
 		if err := h.ejectCdRom(name, namespace, input.DiskNames); err != nil {
```
Comment (aiwantaozi) [inline pkg/api/vm/handler.go]: can we directly return ```h.ejectCdRom(name, namespace, input.DiskNames)```

File: pkg/api/vm/handler.go [hunk 5]
```diff
@@ -27,54 +30,65 @@ type vmActionHandler struct {
...
-			rw.WriteHeader(http.StatusInternalServerError)
-			rw.Write([]byte(err.Error()))
+			return err
+		}
+	case migrate:
+		if err := h.migrate(namespace, name); err != nil {
+			return err
+		}
+	case abortMigration:
+		if err := h.abortMigration(namespace, name); err != nil {
```
Comment (aiwantaozi) [inline pkg/api/vm/handler.go]: same here

File: pkg/api/vm/handler_test.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
+package vm
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pkg/errors"
```
Comment (aiwantaozi) [inline pkg/api/vm/handler_test.go]: move this package "github.com/pkg/errors" to thirdparty packages

Comment (aiwantaozi) [inline pkg/api/vm/handler_test.go:7]: move this package "github.com/pkg/errors" to thirdparty packages

**Comment by gitlawr** [comment]
Updated

**Review by aiwantaozi — APPROVED** [review]

**Pull request merged**